File: Lib/Telerik/Examples/RssReader/RSS Library/RssFeed.vb
Imports System.Collections.Generic
Imports System.Text
Imports System.Xml
Imports System.Xml.Serialization
Imports System.Net
Imports System.IO
Imports Telerik.WinControls
Imports System.Threading

Namespace CS.RSS_Library
    <XmlRoot("rss")> _
    Public Class RssFeed
        <XmlElement("channel")> _
        Public Property Channel() As RssChannel
            Get
                Return m_Channel
            End Get
            Set(value As RssChannel)
                m_Channel = value
            End Set
        End Property
        Private m_Channel As RssChannel

        Public Sub New()
        End Sub

        Public Shared Function Load(url As String) As RssFeed
            Try
                Dim serializer As New XmlSerializer(GetType(RssFeed))
                Using textReader As New XmlTextReader(url)
                    Dim feed As RssFeed = TryCast(serializer.Deserialize(textReader), RssFeed)
                    Return feed
                End Using
            Catch ex As Exception
                Dim feed As RssFeed = LoadAtomFeed(url)
                If feed Is Nothing Then
                    RadMessageBox.Show("Unsupported RSS feed version " + ex.Message)
                End If
                Return feed
            End Try
        End Function

        Private Shared Function LoadAtomFeed(url As String) As RssFeed
            Try
                Dim urlAddress As String = url
                Dim data As String = ""
                Dim request As HttpWebRequest = DirectCast(WebRequest.Create(urlAddress), HttpWebRequest)
                Dim response As HttpWebResponse = DirectCast(request.GetResponse(), HttpWebResponse)
                If response.StatusCode = HttpStatusCode.OK Then
                    Dim receiveStream As Stream = response.GetResponseStream()
                    Dim readStream As StreamReader = Nothing
                    If String.IsNullOrEmpty(response.CharacterSet) Then
                        readStream = New StreamReader(receiveStream)
                    Else
                        readStream = New StreamReader(receiveStream, Encoding.GetEncoding(response.CharacterSet))
                    End If
                    data = readStream.ReadToEnd()

                    response.Close()
                    readStream.Close()
                End If

                Dim xml As New XmlDocument()
                xml.LoadXml(data)

                Dim root As XmlElement = xml.DocumentElement
                If root.Name.ToLower() <> "feed" OrElse Not root.NamespaceURI.Contains("www.w3.org/2005/Atom") Then
                    Return Nothing
                End If

                Dim entries As XmlNodeList = root.GetElementsByTagName("entry")
                Dim atomFeed As New RssFeed()
                Dim channel As New RssChannel()
                channel.Title = root("title").InnerText

                atomFeed.Channel = channel
                Dim rssItems As New List(Of RssItem)()

                For Each entry As XmlElement In entries
                    Dim rssItem As New RssItem()
                    Dim description As String = ""
                    Dim title As String = entry("title").InnerText
                    Dim pubDate As String = entry("published").InnerText
                    Dim link As String = entry("link").Attributes("href").Value

                    If entry("media:group") IsNot Nothing Then
                        If entry("media:group")("media:description") IsNot Nothing Then
                            description = entry("media:group")("media:description").InnerText
                        End If
                        If entry("media:group")("media:thumbnail") IsNot Nothing AndAlso entry("media:group")("media:thumbnail").Attributes("url") IsNot Nothing Then
                            Dim imageUrl As New RssItemImage()
                            imageUrl.Url = entry("media:group")("media:thumbnail").Attributes("url").Value
                            rssItem.ImageUrl = imageUrl
                            Dim width As Integer = 50
                            If entry("media:group")("media:thumbnail").Attributes("width") IsNot Nothing Then
                                width = Integer.Parse(entry("media:group")("media:thumbnail").Attributes("width").Value)
                            End If
                            Dim height As Integer = 50
                            If entry("media:group")("media:thumbnail").Attributes("height") IsNot Nothing Then
                                height = Integer.Parse(entry("media:group")("media:thumbnail").Attributes("height").Value)
                            End If

                            description += "<br>" & "<a href=""" & link & """>" & "<img src=""" & imageUrl.Url & """ width=""" & width & """ height=""" & height & """ border=""0"" >" & "</a>"
                        End If
                    End If

                    If String.IsNullOrEmpty(description) AndAlso entry("content") IsNot Nothing Then
                        description = entry("content").InnerText
                    End If

                    If String.IsNullOrEmpty(description) AndAlso entry("summary") IsNot Nothing Then
                        description = entry("summary").InnerText
                    End If

                    rssItem.Description = description
                    rssItem.Title = title
                    rssItem.PubDate = pubDate
                    rssItem.Link = link
                    rssItems.Add(rssItem)
                Next

                channel.Items = rssItems.ToArray()
                Return atomFeed
            Catch generatedExceptionName As Exception
                Return Nothing
            End Try
        End Function
    End Class
End Namespace

File: Lib/Telerik/Examples/RssReader/RSS Library/RssFeed.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Xml;
using System.Xml.Serialization;
using System.Net;
using System.IO;
using Telerik.WinControls;
using System.Threading;

namespace RssReader.RSS_Library
{
    [XmlRoot("rss")]
    public class RssFeed
    {
        [XmlElement("channel")]
        public RssChannel Channel { get; set; }

        public RssFeed()
        {

        }

        public static RssFeed Load(string url)
        {
            try
            {
                XmlSerializer serializer = new XmlSerializer(typeof(RssFeed));
                using (XmlTextReader textReader = new XmlTextReader(url))
                {
                    RssFeed feed = serializer.Deserialize(textReader) as RssFeed;
                    return feed;
                }
            }
            catch (Exception ex)
            {
                RssFeed feed = LoadAtomFeed(url);
                if (feed == null)
                {
                    RadMessageBox.Show("Unsupported RSS feed version " + ex.Message);
                }
                return feed;
            }
        }

        private static RssFeed LoadAtomFeed(string url)
        {
            try
            {
                string urlAddress = url;
                string data = "";
                HttpWebRequest request = (HttpWebRequest)WebRequest.Create(urlAddress);
                HttpWebResponse response = (HttpWebResponse)request.GetResponse();
                if (response.StatusCode == HttpStatusCode.OK)
                {
                    Stream receiveStream = response.GetResponseStream();
                    StreamReader readStream = null;
                    if (string.IsNullOrEmpty(response.CharacterSet))
                        readStream = new StreamReader(receiveStream);
                    else
                        readStream = new StreamReader(receiveStream, Encoding.GetEncoding(response.CharacterSet));
                    data = readStream.ReadToEnd();

                    response.Close();
                    readStream.Close();
                }

                XmlDocument xml = new XmlDocument();
                xml.LoadXml(data);

                XmlElement root = xml.DocumentElement;
                if (root.Name.ToLower() != "feed" || !root.NamespaceURI.Contains("www.w3.org/2005/Atom"))
                {
                    return null;
                }

                XmlNodeList entries = root.GetElementsByTagName("entry");
                RssFeed atomFeed = new RssFeed();
                RssChannel channel = new RssChannel();
                channel.Title = root["title"].InnerText;

                atomFeed.Channel = channel;
                List<RssItem> rssItems = new List<RssItem>();

                foreach (XmlElement entry in entries)
                {
                    RssItem rssItem = new RssItem();
                    string description = "";
                    string title = entry["title"].InnerText;
                    string pubDate = entry["published"].InnerText;
                    string link = entry["link"].Attributes["href"].Value;

                    if (entry["media:group"] != null)
                    {
                        if (entry["media:group"]["media:description"] != null)
                        {
                            description = entry["media:group"]["media:description"].InnerText;
                        }
                        if(entry["media:group"]["media:thumbnail"] != null &&
                            entry["media:group"]["media:thumbnail"].Attributes["url"] != null)
                        {
                            RssItemImage imageUrl = new RssItemImage();
                            imageUrl.Url = entry["media:group"]["media:thumbnail"].Attributes["url"].Value;
                            rssItem.ImageUrl = imageUrl;
                            int width = 50;
                            if (entry["media:group"]["media:thumbnail"].Attributes["width"] != null)
                            {
                                width = int.Parse(entry["media:group"]["media:thumbnail"].Attributes["width"].Value);
                            }
                            int height = 50;
                            if (entry["media:group"]["media:thumbnail"].Attributes["height"] != null)
                            {
                                height = int.Parse(entry["media:group"]["media:thumbnail"].Attributes["height"].Value);
                            }

                            description += "<br>" +
                                            "<a href=\"" + link + "\">" +
                                                "<img src=\"" + imageUrl.Url + "\" width=\"" + width + "\" height=\"" + height + "\" border=\"0\" >" +
                                            "</a>";
                        }
                    }

                    if (string.IsNullOrEmpty(description) && entry["content"] != null)
                    {
                        description = entry["content"].InnerText;
                    }

                    if (string.IsNullOrEmpty(description) && entry["summary"] != null)
                    {
                        description = entry["summary"].InnerText;
                    }

                    rssItem.Description = description;
                    rssItem.Title = title;
                    rssItem.PubDate = pubDate;
                    rssItem.Link = link;
                    rssItems.Add(rssItem);
                }

                channel.Items = rssItems.ToArray();
                return atomFeed;
            }
            catch (Exception)
            {
                return null;
            }
        }
    }
}
